### ISSUE TITLE
add option to `hub fork` to set the name of the remote that is created

### ISSUE THREAD

**Issue by ryan-williams** [github/hub#762]
it creates a remote named `$GITHUB_USER` but I frequently want to use something shorter ("ryan" vs. "ryan-williams")

**Comment by jingweno** [comment]
Duplicated of https://github.com/github/hub/issues/754

**Comment by ryan-williams** [comment]
It seems like #697 wants the fork on github to have a different name; I just want the created remote (that points to the fork) in my local git repo to have a name other than my `$GITHUB_USER`.

The former seems a good deal trickier than the latter, which is what I was asking for here, so I'm not sure I agree that this is a duplicate of that, fwiw.

**Comment by ryan-williams** [comment]
Let me know if I'm misreading or misunderstanding.

**Comment by jingweno** [comment]
Sorry, maybe not completely duplicate. But could be solved if https://github.com/github/hub/issues/697 is implemented - just provide your own fork origin name.

**Comment by ryan-williams** [comment]
maybe, but actually the name of the fork and the name of my local git-remote that points at it are two distinct things.

I personally always like my forks to have the same name as the upstream, so I have no interest in #697 being solved (if I'm correctly understanding what's being asked for there).

I just don't like to use my github username to name my git-remotes that point at my forks.

**Comment by ryan-williams** [comment]
sorry to be pedantic, but it seems like this could also be solved much more easily than #697, and I don't think their respective solutions would have much to do with one another

**Comment by ryan-williams** [comment]
(having not actually looked at the implementation of `hub fork`, just surmising from these issues and the way I would guess that it all works)

**Comment by jingweno** [comment]
> maybe, but actually the name of the fork and the name of my local git-remote that points at it are two distinct things.

The remote name of your fork repository is what I meant. I'm thinking something like this:

```
$ git clone github/hub hub-custom
$ cd hub-custom
$ hub fork ryan-williams
```

What do you think?

**Comment by jingweno** [comment]
Ah...sorry, I misunderstood what https://github.com/github/hub/issues/697 does. I thought it only means the remote name...reopening

**Comment by jingweno** [comment]
https://github.com/github/hub/issues/754 is a closer issue

**Comment by ryan-williams** [comment]
interesting, yea #754 is more similar, though it has different ideas for both what the original repo should be called and what the fork should be called, maybe what the original should be called **from the fork**? I guess making all of those things configurable could be nice. Seems like we're getting into the weeds of peoples' personal preferences around this stuff, but that in itself is probably motivation for making some or all of it more configurable.

Thanks for the pointers and for re-opening!

**Comment by brock** [comment]
I started the thread at https://github.com/isaacs/github/issues/280, and at @kbrock's suggestion I began revisiting `hub` based on that thread. Here were my observations:

    `hub fork` is run from inside a clone.

This wasn't exactly what I was expecting, and might be why this issue is so challenging. What I was expecting was this:

```
    $ hub fork github/hub
    $ git remote show
    origin --> brock/hub
    upstream --> github/hub
```

It wouldn't occur to me to think of forking a repo from inside a cloned repo, but I'll admit that my perspective is one of a complete outsider, so don't throw me under the bus, just wanted to share my thoughts from someone outside looking in.

Would guys be open to a discussion around an update to the `hub fork` command so that it checks for arguments passed and uses the above behavior in that scenario?

**Comment by kbrock** [comment]
I would think adding a git alias called `fork-name`(?) would allow people to customize it how ever they wanted. You have access to bash, remote names and everything. Could put it into a global `.gitconfig`, or put it into the path with the name `git-fork-name`.

Per @brock's comment, I often know I want to fork first. I also like `origin` to point to `brock` and `upstream` to point to `github`.

**Comment by mislav** [comment]
I kept track of features that you people are suggesting. Could someone summarize it for me?

**Comment by kbrock** [comment]
@ryan-williams is the current consensus?

```
$ git clone github/hub
$ cd hub
$ hub fork ryan-williams ryan
$ git remote -v
ryan git@github.com/ryan-williams/hub.git
```

**Comment by brock** [comment]
| you people

(leaves thread) this feature isn't that important to me

**Comment by ryan-williams** [comment]
@kbrock that seems reasonable, though I'd vote for a little magical-default-intelligence too, e.g. in your example above:
* must `ryan-williams` (my github username) be specified if I already have `$GITHUB_USER` set?
* if I always want my [local git-remote pointing to my fork] to be called `ryan`, can that be an environment- or git-config-variable too? `$HUB_FORK_NAME`?

that way, I could just say:
```
$ git clone github/hub
$ cd hub
$ hub fork  # <-- win!
$ git remote -v
ryan git@github.com/ryan-williams/hub.git  # this is how I always want it to look when I fork things.
```

My guess is that everyone has a certain name that they always use for their [local git-remote pointing to their github fork] to be called. In my case it's "ryan" and is always "ryan"; it sounds like for other people it's actually "origin" (and the thing I call origin they call "upstream"). I don't think either of these is necessarily better than the other, but the question is whether for each person it is always the same every time they create a fork, in which case magical-defaults could be nice!

**Comment by irthos** [comment]
fwiw, I'd like to programmatically fork a repo to my account without first cloning it on my machine.

**Comment by mislav** [comment]
@irthos I hear you. But that is a separate feature than the one suggested in this thread.

**Comment by irthos** [comment]
@mislav makes sense. I'll see if it's a dupe request or post it alone.

**Comment by kbrock** [comment]
@irthos great idea - with that, it would be easy to script a fork command in git to do anything you want

**Comment by corysimmons** [comment]
Any updates on this?

I like https://github.com/github/hub/issues/762#issuecomment-74908996 by default and https://github.com/github/hub/issues/762#issuecomment-74868642 (although I'd swap remote and username positions so people can set default username and set remote names on the fly) to override.

**Comment by joshuajabbour** [comment]
Since none of these are required, shouldn't they be flags and not arguments? What about `hub fork -o [REMOTE] [-u] [USER]`? The `-o` isn't ideal as a descriptive flag, but matches `git clone`. The `-u` follows the pattern set by other hub commands.

**Comment by un1versal** [comment]
The **origin** (unforked original project) should be named **origin** no matter what
**forks** should be named **fork** (not username, not something else) name it what it is.

when you fork something then clone it, the origin should become automatically upstream in the worst case scenario. you your fork still be named **FORK**

Dont Fork this up anymore than already is... #768

**Comment by mislav** [comment]
Available in hub v2.3.0 8a4d8dd0eb36954aa9fd4b2c70d4796299730df8

**Comment by kbrock** [comment]
Thank you for your help @mislav
